lib/Controller/DirectUploadController.php

```
@@ -104,10 +126,107 @@ public function prepareDirectUpload(int $folder_id):DataResponse {
 					'error' => 'folder not found or not enough permissions'
 				], Http::STATUS_NOT_FOUND);
 			}
-		} catch (Exception $e) {
+		} catch (Exception | NotPermittedException |NoUserException $e) {
 			return new DataResponse([
 				'error' => 'folder not found or not enough permissions'
 			], Http::STATUS_NOT_FOUND);
 		}
 	}
+
+	/**
+	 * preparation for the direct upload
+	 *
+	 * @CORS
+	 * @NoCSRFRequired
+	 * @NoAdminRequired
+	 * @PublicPage
+	 *
+	 * This can be tested with:
+	 * curl -X POST http://my.nc.org/index.php/apps/integration_openproject/direct-upload/<token>
+	 *
+	 * @param string $token
+	 * @param string $file_name
+	 * @param string $contents
+	 *
+	 * @return DataResponse
+	 */
+	public function directUpload(string $token, string $file_name, string $contents):DataResponse {
+		$fileId = null;
+		try {
+			if (strlen($token) !== 64 || !preg_match('/^[a-zA-Z0-9]*/', $token)) {
+				return new DataResponse([
+					'error' => 'Invalid token. The token should be 64 characters long and in human readable format.'
```

not sure if we want to tell anyone who is guessing tokens what the format should be. Use the token that you generated, if you send anything else you will receive a `'Invalid token'` message.
That check could also be integrated into `getTokenInfo()` or you could throw an exception here `throw new NotFoundException('Invalid token.');` then there will be just one place where the DataResponse is generated

---

I didn't think it that way. I'll refactor it